exporter/googlecloudpubsubexporter/factory.go

```diff
@@ -0,0 +1,129 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package googlecloudpubsubexporter
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/exporter/exporterhelper"
+)
+
+const (
+	// The value of "type" key in configuration.
+	typeStr        = "googlecloudpubsub"
+	defaultTimeout = 12 * time.Second
+)
+
+func NewFactory() component.ExporterFactory {
+	return exporterhelper.NewFactory(
+		typeStr,
+		createDefaultConfig,
+		exporterhelper.WithTraces(createTracesExporter),
+		exporterhelper.WithMetrics(createMetricsExporter),
+		exporterhelper.WithLogs(createLogsExporter))
+}
+
+var exporters = map[config.Exporter]*pubsubExporter{}
```

Comment:
Could you add some comments explaining why this global mutable state is necessary? Based on a spot check, I didn't see such a pattern used in other exporters.

Reply:
A lot of exporter/receivers have other endpoints to send data to/from. Here everything is piped over a single pubsub topic/subscription so there is a need to reuse the same connection.

Reply:
```suggestion
// This map allows multiple exporter instances to be multiplexed over the same PubSub connection.
// It is initialized at process startup and read-only afterwards, so we don't need to guard it for concurrent access.
var exporters = map[*Config]*pubsubExporter{}
```

I also suggest changing the map key type to the concrete type that we care about rather than an interface, because that documents the intent better.

Reply:
1. It originally was the concrete type but making it the interface removed the need to cast at that point
2. Yes, the exporter joins the same topic, so the same gRPC connection to Pubsub

Note: this rule of splitting a PR into different phases makes it really annoying to contribute a new module. I have the full implementation working for more than half a year, but now have to retrofit all the comments back in that implementation hoping I don't do mistakes, and then merge that back for PR2 (the implementation). Having the implementation already available at the initial commit would make reviewing also easier. Just my 2 cents.